common/src/primitives/merkle/proof/single/mod.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2021-2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::primitives::H256;
+
+use super::super::{
+    tree::{MerkleTree, Node},
+    MerkleTreeProofExtractionError,
+};
+
+/// A proof for a single leaf in a Merkle tree. The proof contains the leaf and the branch of the tree.
+/// This is considered an intermediary object. For storage, use `SingleProofHashes` through
+/// `SingleProofNodes::into_values()`.
+#[must_use]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SingleProofNodes<'a> {
+    leaf: Node<'a>,
+    branch: Vec<Node<'a>>,
+}
+
+impl<'a> SingleProofNodes<'a> {
+    /// Creates a proof for a leaf by its index in the lowest level (the tip).
+    /// A proof doesn't contain the root.
+    pub fn from_tree_leaf(
+        tree: &'a MerkleTree,
+        leaf_index: usize,
+    ) -> Result<Self, MerkleTreeProofExtractionError> {
+        let leaf_count = tree.leaf_count().get();
+        if leaf_index > leaf_count {
+            return Err(MerkleTreeProofExtractionError::LeafIndexOutOfRange(
+                leaf_index, leaf_count,
+            ));
+        }
+
+        let leaf = tree.node_from_bottom(0, leaf_index).ok_or(
+            MerkleTreeProofExtractionError::AccessError(
+                crate::primitives::merkle::MerkleTreeAccessError::AbsIndexOutOfRange(
+                    leaf_index,
+                    tree.total_node_count().get(),
+                ),
+            ),
+        )?;
+
+        let mut proof = vec![];
+
+        // once we reach root we stop
+        for node in leaf.into_iter_parents() {
+            if node.is_root() {
+                break;
+            }
+            // We push siblings of parents because they're what we need to calculate the root, upwards.
+            let err_msg = "In this loop, this cannot be root, so sibling must exist";
+            let sibling = node.sibling().expect(err_msg);
+            proof.push(sibling);
+        }
```

Comment:
How about something like:
```rust
let proof: Vec<_> = leaf.into_iter_parents().map_while(Node::sibling).collect();
```

Reply:
Done. Though I added an assert to ensure the size makes sense.